database/database.go

```diff
@@ -1,36 +1,46 @@
-package nosql
+package database
 
 import (
-	"errors"
 	"fmt"
+
+	"errors"
 )
 
 var (
 	// ErrNotFound is the type returned on DB implementations if an item does not
 	// exist.
 	ErrNotFound = errors.New("not found")
+	// ErrOpNotSupported is the type returned on DB implementations if an operation
+	// is not supported.
+	ErrOpNotSupported = errors.New("operation not supported")
 )
 
+// Options are configuration options for the database.
+type Options struct {
+	Database string
+	ValueDir string
+}
+
```

Comment:
type Option as a `func(typ string, o options) error` will be more future proof.

For example:
```go
type options struct {
    badger *bagder.Options
}
func WithValueDir(name string) Options {
     return func (typ string, o *options) {
          if typ == "badger" {
              o.badger.ValueDir = name
          }
          return nil
     }
}
```
or instead with `func (interface{}) error` and do type switches so we have the database type implicit.

Reply:
and `Open(dataSourceName string, options ...Option)`

Reply:
Not that I think a problem with that is that a program compiled with that will have all types compiled in the binary even if they are not using it ...

A way to solve this is to have the `WithValueDir` as a method in the badger package.

It's something to think about, right now you can just set the fixed string in options with
```go
type Options struct{
	Database string
	ValueDir string
}

type Option func(opt *Options) error

func WithValueDir(name string) Option {
	return func(o *Options) error {
		o.ValueDir = name
		return nil
	}
}

func WithDatabase(name string) Option {
	return func(o *Options) error {
		o.Database = name
		return nil
	}
}
```